### PR TITLE
Additional documentation and tests related to SourceLocation

### PR DESCRIPTION
**Pull request by aaronhurst-google**

### CODE REVIEW

File: src/main/java/com/google/summit/ast/SourceLocation.kt [hunk 1]
```diff
@@ -21,6 +21,11 @@ package com.google.summit.ast
  *
  * A special "unknown" value can also be constructed.
  *
+ * The range is inclusive of the end character.
+ * The line numbering begins at 1.
+ * The column numbering begins at 0.
```
Comment (zorzella) [inline src/main/java/com/google/summit/ast/SourceLocation.kt]: I never thought carefully about this: is this lack of symmetry "standard"? And/or are you going with this choice for compatibility with some other layer (antlr, pmd)?

File: src/main/java/com/google/summit/ast/SourceLocation.kt [hunk 2]
```diff
@@ -21,6 +21,11 @@ package com.google.summit.ast
...
+ * On empty lines, the column is 0.
```
Comment (zorzella) [inline src/main/java/com/google/summit/ast/SourceLocation.kt]: I'm not sure I understand this sentence, maybe related to a question that's not quite about this change itself, but about the [endColumn]: is that "inclusive" or "exclusive"?

File: src/main/java/com/google/summit/translation/Translate.kt [hunk 3]
```diff
@@ -1350,9 +1351,11 @@ class Translate(val file: String, private val tokens: TokenStream) : ApexParserB
       *visitSoslClauses(ctx.soslClauses()).bindings.toTypedArray(),
       ctx.boundExpression()?.let { visitBoundExpression(it) },
     ).filterNotNull()
-    return SoslExpression(toSourceString(ctx),
-                          bindings,
-                          toSourceLocation(ctx))
+    return SoslExpression(
+        toSourceString(ctx).trim().removeSurrounding("[","]"),
```
Comment (zorzella) [inline src/main/java/com/google/summit/translation/Translate.kt]: this change looks unrelated, no?

File: src/main/javatests/com/google/summit/translation/SourceLocationTest.kt [hunk 4]
```diff
@@ -116,12 +115,38 @@ class SourceLocationTest {
...
+  fun trailingEmptyLine() {
+    val input = "public class Test { }\n"
+
+    val cu = TranslateHelpers.parseAndTranslate(input)
+    val loc = cu.getSourceLocation()
+
+    assertThat(loc.endLine).isEqualTo(2)
+    // There is no first column on the second, empty line.
+    // Expect zero in this case.
+    assertThat(loc.endColumn).isEqualTo(0)
```
Comment (zorzella) [inline src/main/javatests/com/google/summit/translation/SourceLocationTest.kt:135]: I guess this test clarifies to me a bit what you meant with the comment I ask above. Maybe I'm still missing some subtlety here... what happens if:

```
val input = "public class Test { }\n "

or

val input = "public class Test { }\n  "
```

?

File: src/main/javatests/com/google/summit/translation/SourceLocationTest.kt [hunk 5]
```diff
@@ -116,12 +115,38 @@ class SourceLocationTest {
...
+  }
+
+  @Test
+  fun leadingEmptyLine() {
+    val input = "\npublic class Test { }"
+
+    val cu = TranslateHelpers.parseAndTranslate(input)
+    val loc = cu.getSourceLocation()
+
+    // The source location starts from the first regular token.
```
Comment (zorzella) [inline src/main/javatests/com/google/summit/translation/SourceLocationTest.kt:158]: why this lack of symmetry? And what happens if

val input = "\n public class Test { }"

or

val input = " public class Test { }"

?

File: src/main/java/com/google/summit/ast/SourceLocation.kt (hunk 1 above)
Comment (aaronhurst-google) [reply src/main/java/com/google/summit/ast/SourceLocation.kt]: Not sure, but ANTLR and PMD disagree. Summit copied the behavior of ANTLR. I'm documenting the existing behavior.

https://www.antlr.org/api/Java/org/antlr/v4/runtime/Token.html#getCharPositionInLine()

File: src/main/java/com/google/summit/ast/SourceLocation.kt (hunk 2 above)
Comment (aaronhurst-google) [reply src/main/java/com/google/summit/ast/SourceLocation.kt]: It was exclusive, but it is now inclusive. This is was the primary point of this change.

This matches PMD... and human intuition (I think).

File: src/main/java/com/google/summit/translation/Translate.kt (hunk 3 above)
Comment (aaronhurst-google) [reply src/main/java/com/google/summit/translation/Translate.kt]: Errr, sorry, this is a bit from the other PR #28 from earlier this week.

(P.S. That PR needs to be reapproved. It lost the approval bit after I made the requested changes.)

File: src/main/javatests/com/google/summit/translation/SourceLocationTest.kt (hunk 4 above)
Comment (aaronhurst-google) [reply src/main/javatests/com/google/summit/translation/SourceLocationTest.kt:135]: The test is testing when the end position is on an empty line because: what should the `endColumn` be when there are no columns? (I chose zero.)

That ambiguity goes away with your two cases. For #1, the `endColumn=0` will include all of the input characters. For #2, the `endColumn=1`.

File: src/main/javatests/com/google/summit/translation/SourceLocationTest.kt (hunk 5 above)
Comment (aaronhurst-google) [reply src/main/javatests/com/google/summit/translation/SourceLocationTest.kt:158]: It follows from the grammar, not from my code that's representing source locations. No great reason. I threw in this test to see if there was also a degenerate case with `beginColumn` on an empty line, but I don't think it can happen with this grammar.

Comment (aaronhurst-google) [reply src/main/javatests/com/google/summit/translation/SourceLocationTest.kt:158]: (All of these cases have a `beginColumn` at the `p` character.)

File: src/main/javatests/com/google/summit/translation/SourceLocationTest.kt (hunk 4 above)
Comment (aaronhurst-google) [reply src/main/javatests/com/google/summit/translation/SourceLocationTest.kt:135]: I added both tests.

File: src/main/java/com/google/summit/translation/Translate.kt (hunk 3 above)
Comment (aaronhurst-google) [reply src/main/java/com/google/summit/translation/Translate.kt]: Cleaned up.

File: src/main/java/com/google/summit/ast/SourceLocation.kt (hunk 2 above)
Comment (adangel) [reply src/main/java/com/google/summit/ast/SourceLocation.kt]: In PMD 6, I think we use endColumn to be inclusive (for what ever reason, it was always inclusive). Also, for whatever reason, it's zero-based.

In PMD 7, this will change (see [FileLocation](https://github.com/pmd/pmd/blob/db966c574cccc82186a1d0f2b0ebcc12615319d6/pmd-core/src/main/java/net/sourceforge/pmd/lang/document/FileLocation.java#L78-L96)): columns are always 1-based (as are lines), if there is some offset, that will be 0-based. This more represents what you see in all text editors, which display the cursor position.

That's a subtle change to keep in mind, when we bring this apex change to PMD 7.

Comment (zorzella) [reply src/main/java/com/google/summit/ast/SourceLocation.kt]: FWIW "range-like" APIs generally use inclusive starts and exclusive ends (e.g. https://docs.oracle.com/javase/7/docs/api/java/lang/String.html#substring(int,%20int)) for a variety of reasons, including:

a. it's easiest to make adjacent ranges -- e.g. [1, 3) and [3, 7), rather than the equivalent [1, 2], [3, 6]

b. for non-discreet types, it's conceptually impossible to create adjacent ranges. E.g. imagine the numbers are real numbers (say something like Java's BigDecimal), rather than integers...

c. Even if we have fixed point decimals, the whole thing is inconvenient -- e.g. for a 4 decimal points, it'd be something like [1, 2.9999], [3, 6.9999].

d. And don't even get me started with the headaches that involve date ranges, if they are modeled as inclusive/exclusive...

e. I know, I know, this here API deals with integers, so "b", "c" and "d" don't quite apply -- but the point is that because of things like these, as I said, range-like APIs are *generally* designed with this inclusive/exclusive paradigm

f. with inc/exc, it's natural to represent empty intervals. E.g. [10, 10) is perfectly readable. There's no good (non-hacky) way to do this with both inclusive. FWIW IMHO this is manifesting here in the form of your note about empty lines...

If representing internally like this is really a problem for PMD, I don't really mind this change, but I want to be sure we do this quite deliberately... Maybe we can chat about this a bit.

Comment (aaronhurst-google) [reply src/main/java/com/google/summit/ast/SourceLocation.kt]: One other bit of food for thought: this situation happens to concern columns, but what about line ranges? Aren't these pretty universally inclusive ranges? (At least I've never been shown an exclusive range by a compiler, editor, etc.)

I went looking for other implementations of source ranges:
* PMD is inclusive of endpoint (as we know)
* LSP is exclusive of endpoint [spec](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#range)
* Eclipse ranges are length-specified [source](https://github.com/eclipse/aspectj.eclipse.jdt.core/blob/a05312e746b9bc2b48b4b039f6e7b5e061b5b393/org.eclipse.jdt.core/model/org/eclipse/jdt/core/SourceRange.java)
* Semgrep is inclusive of endpoint [source](https://github.com/returntocorp/semgrep/blob/73381198919db13899689f74107bb6b359193df0/semgrep-core/src/core/Range.mli#L6)
* SARIF format is exclusive of endpoint [spec](http://docs.oasis-open.org/sarif/sarif/v2.0/csprd01/sarif-v2.0-csprd01.html#_Toc517436108)

Comment (aaronhurst-google) [reply src/main/java/com/google/summit/ast/SourceLocation.kt]: To summarize: the resolution is to keep the ranges as exclusive of the final character. We can translate this into the values expected by PMD.

**Review by zorzella — APPROVED** [review]
